nightfall/api.py

```diff
@@ -1,153 +1,261 @@
+        if not uploaded:
+            raise NightfallSystemError("File upload failed", 50000)
 
-            self.logger.debug(f"HTTP Status Code: {response.status_code}")
-            self.logger.debug(f"HTTP Response Headers: {response.headers}")
-            self.logger.debug(f"HTTP Response Text: {response.text}")
+        response = self._file_scan_finalize(session_id)
+        _validate_response(response, 200)
 
-            response.raise_for_status()
-            findings = response.json()
+        response = self._file_scan_scan(session_id, webhook_url,
+                                        policy_uuid=policy_uuid,
+                                        detection_rule_uuids=detection_rule_uuids,
+                                        detection_rules=detection_rules)
+        _validate_response(response, 200)
+        parsed_response = response.json()
 
-            for idx, d in enumerate(chunk):
-                if findings[idx] is not None:
-                    all_findings[d] = findings[idx]
-                else:
-                    all_findings[d] = None
+        return parsed_response["id"], parsed_response["message"]
+
+    def _file_scan_initialize(self, location: str):
+        data = {
+            "fileSizeBytes": os.path.getsize(location)
+        }
+        response = requests.post(
+            url=self.FILE_SCAN_INITIALIZE_ENDPOINT,
+            headers=self._headers,
+            data=json.dumps(data)
+        )
+
+        return response
+
+    def _file_scan_upload(self, session_id: str, location: str, chunk_size: int):
+
+        def read_chunks(fp, chunk_size):
+            ix = 0
+            while True:
+                data = fp.read(chunk_size)
+                if not data:
+                    break
+                yield ix, data
+                ix = ix + 1
+
+        def upload_chunk(id, data, headers):
+            response = requests.patch(
+                url=self.FILE_SCAN_UPLOAD_ENDPOINT.format(id),
+                data=data,
+                headers=headers
+            )
+            return response
+
+        with open(location) as fp:
+            for ix, piece in read_chunks(fp, chunk_size):
+                headers = self._headers
+                headers["X-UPLOAD-OFFSET"] = str(ix * chunk_size)
+                response = upload_chunk(session_id, piece, headers)
+                _validate_response(response, 204)
+
+        return True
+
+    def _file_scan_finalize(self, session_id: str):
+        response = requests.post(
+            url=self.FILE_SCAN_COMPLETE_ENDPOINT.format(session_id),
+            headers=self._headers
+        )
+        return response
+
+    def _file_scan_scan(self, session_id: str, webhook_url: str, policy_uuid: str,
+                        detection_rule_uuids: List[str], detection_rules: List[DetectionRule]):
+        if policy_uuid:
+            data = {"policyUUID": policy_uuid}
```

Comment:
Should webhook URL be optional if it's not used for when `policy_uuid` is supplied?